crates/aiken-lang/src/tests/check.rs

```diff
@@ -1779,15 +1779,143 @@ fn backpassing_type_annotation() {
             transition_fold4(
               x,
             )
-          
+
           fn(g){
             g(if input.foo == 1{
               1
             } else {
               2
             })
           }
-          
+
+        }
+    "#;
+
+    assert!(check(parse(source_code)).is_ok())
+}
+
+#[test]
+fn forbid_expect_into_opaque_type_from_data() {
+    let source_code = r#"
+        opaque type Thing { inner: Int }
+
+        fn bar(n: Data) {
+          expect a: Thing = n
+
+          a
+        }
+    "#;
+
+    assert!(matches!(
+        check(parse(source_code)),
+        Err((_, Error::ExpectOnOpaqueType { .. }))
+    ))
+}
+
+#[test]
+fn forbid_expect_into_opaque_type_constructor_without_typecasting() {
+    let source_code = r#"
+        opaque type Thing {
+          Foo(Int)
+          Bar(Int)
+        }
+
+        fn bar(thing: Thing) {
+          expect Foo(a) = thing
+          a
+        }
+    "#;
+
+    assert!(matches!(
+        check(parse(source_code)),
+        Err((_, Error::ExpectOnOpaqueType { .. }))
+    ))
+}
+
+#[test]
+fn forbid_expect_into_opaque_type_constructor_with_typecasting() {
+    let source_code = r#"
+        opaque type Thing {
+          Foo(Int)
+          Bar(Int)
+        }
+
+        fn bar(data: Data) {
+          expect Foo(a): Thing = data
+          a
+        }
+    "#;
+
+    assert!(matches!(
+        check(parse(source_code)),
+        Err((_, Error::ExpectOnOpaqueType { .. }))
+    ))
+}
+
+#[test]
+fn forbid_expect_into_nested_opaque_in_record_without_typecasting() {
+    let source_code = r#"
+        opaque type Thing { inner: Int }
+
+        type Foo { foo: Thing }
+
+        fn bar(thing: Foo) {
+          expect Foo { foo: Thing { inner } } : Foo = thing
+          Void
+        }
+    "#;
+
+    assert!(matches!(
+        check(parse(source_code)),
+        Err((_, Error::ExpectOnOpaqueType { .. }))
+    ))
+}
+
+#[test]
+fn forbid_expect_into_nested_opaque_in_record_with_typecasting() {
+    let source_code = r#"
+        opaque type Thing { inner: Int }
+
+        type Foo { foo: Thing }
+
+        fn bar(a: Data) {
+          expect Foo { foo: Thing { inner } } : Foo = a
+          Void
+        }
+    "#;
+
+    assert!(matches!(
+        check(parse(source_code)),
+        Err((_, Error::ExpectOnOpaqueType { .. }))
+    ))
+}
+
+#[test]
+fn forbid_expect_into_nested_opaque_in_list() {
+    let source_code = r#"
+        opaque type Thing { inner: Int }
+
+        fn bar(a: Data) {
+          expect [x]: List<Thing> = [a]
+          Void
+        }
+    "#;
+
+    assert!(matches!(
+        check(parse(source_code)),
+        Err((_, Error::ExpectOnOpaqueType { .. }))
+    ))
+}
+
+#[test]
+fn allow_expect_on_var_patterns_that_are_opaque() {
+    let source_code = r#"
+        opaque type Thing { inner: Int }
+
+        fn bar(a: Option<Thing>) {
+          expect Some(thing) = a
+
+          thing.inner
```

Comment:
Crap indeed. So we do need to relax the check to allow expect if:

1. rhs doesn't type-cast into an opaque  (i.e. not turning a `Data` into an opaque type).
2. lhs pattern do not use opaque type constructors

To make things _simpler_,  I think that (1) can simply error is there's any `Data` on the rhs. In principle, we should only error if the cast happens on an opaque type but this might be a lot trickier to get right. For example, a list of `List<(Opaque, Data)>` could in theory be turned into a `List<(Opaque, NotOpaque)>`  without restrictions. But detecting that sounds a lot harder. The use-case seems quite limited and not required at the moment. 

I think also we get (2) "for free" due to how type constructors are registered and scoped by module. Opaque constructors simply do not exist outside of the module that defines the opaque type, and thus, shouldn't be available in a pattern to begin with.